### PR TITLE
Add topic, service, and action type information to ROS2State

### PR DESCRIPTION
**Pull request by schmerl**
The type information in ROS2 can be gleaned from ros2 info, and so we will cache it in the state.

### CODE REVIEW

File: src/roswire/ros2/state.py [hunk 1]
```diff
@@ -41,6 +41,12 @@ class ROS2SystemState(SystemState):
         A mapping from actions to the names of providesr of that action.
     action_clients: Mapping[str, Collection[str]]
         A mapping from actions to the names of clients of that action
+    topic_to_type: Mapping[str, str]
+        A mapping from topics to their type
+    service_to_type: Mapping[str, str]
+        A mapping from services to their type
+    action_to_type:
```
Comment (ChrisTimperley) [inline src/roswire/ros2/state.py:48]: Missing type annotation

File: src/roswire/ros2/state.py [hunk 2]
```diff
@@ -41,6 +41,12 @@ class ROS2SystemState(SystemState):
         A mapping from actions to the names of providesr of that action.
     action_clients: Mapping[str, Collection[str]]
         A mapping from actions to the names of clients of that action
+    topic_to_type: Mapping[str, str]
+        A mapping from topics to their type
```
Comment (ChrisTimperley) [inline src/roswire/ros2/state.py:45]: For clarity, we should change `to their type` to `to the name of their type`

File: src/roswire/ros2/state.py [hunk 3]
```diff
@@ -116,6 +125,11 @@ def probe(self) -> ROS2SystemState:
             _ACTION_SERVERS: {},
             _ACTION_CLIENTS: {}
         }
+        # The place to store type information
```
Comment (ChrisTimperley) [inline src/roswire/ros2/state.py:128]: I don't think that the comment is really necessary here

File: src/roswire/ros2/state.py [hunk 4]
```diff
@@ -139,37 +154,57 @@ def probe(self) -> ROS2SystemState:
...
                     if name in node_to_state[mode]:
                         node_to_state[mode][name].add(node_name)
                     else:
                         node_to_state[mode][name] = {node_name}
+                    # Add type information and report conflict if a
+                    # different types was registered (probably
+                    # should never happen)
+                    if name in types and fmt != types[name]:
+                        logger.warning(
+                            f'The entity {name} has conflictig types: '
```
Comment (ChrisTimperley) [inline src/roswire/ros2/state.py:193]: typo

File: src/roswire/ros2/state.py [hunk 5]
```diff
@@ -139,37 +154,57 @@ def probe(self) -> ROS2SystemState:
...
+                    name = partition[0]
+                    fmt = partition[1]
                     if name in node_to_state[mode]:
                         node_to_state[mode][name].add(node_name)
                     else:
                         node_to_state[mode][name] = {node_name}
+                    # Add type information and report conflict if a
+                    # different types was registered (probably
+                    # should never happen)
+                    if name in types and fmt != types[name]:
```
Comment (ChrisTimperley) [inline src/roswire/ros2/state.py:191]: This is a nice sanity check, but I think we should either raise an exception or use an assertion here. (Continuing with the execution could lead to strange results.)

Comment (schmerl) [reply src/roswire/ros2/state.py:198]: Chose to raise an exception.

File: src/roswire/ros2/state.py (hunk 1 above)
Comment (schmerl) [reply src/roswire/ros2/state.py]: Done

File: src/roswire/ros2/state.py (hunk 2 above)
Comment (schmerl) [reply src/roswire/ros2/state.py]: Done

File: src/roswire/ros2/state.py (hunk 3 above)
Comment (schmerl) [reply src/roswire/ros2/state.py]: OK.

File: src/roswire/ros2/state.py (hunk 4 above)
Comment (schmerl) [reply src/roswire/ros2/state.py]: Done,

File: src/roswire/ros2/state.py (hunk 5 above)
Comment (schmerl) [reply src/roswire/ros2/state.py:198]: Raising an exception

File: src/roswire/exceptions.py [hunk 6]
```diff
@@ -158,6 +158,19 @@ class ParsingError(ROSWireException):
...
+@_attr.s(frozen=True, auto_exc=True, auto_attribs=True, str=False)
+class ConflictingTypes(ROSWireException):
+    """ROSWire found two or more types for a service, topic, or action."""
+
+    entity: str
+    existing: str
+    conflicting: str
+
+    def __str__(self) -> str:
+        return f"error declaring type {self.conflicting} for {self.entity}. " \
```
Comment (ChrisTimperley) [inline src/roswire/exceptions.py:170]: Prefer the use of parens to break up strings over `\`.
E.g.,
```
("foo"
 "bar")
```

File: src/roswire/ros2/state.py [hunk 7]
```diff
@@ -106,7 +116,13 @@ def for_app_instance(cls, app_instance: "AppInstance") -> "ROS2StateProbe":
         return ROS2StateProbe(app_instance=app_instance)
 
     def probe(self) -> ROS2SystemState:
-        """Obtains the instantaneous state of the associated ROS system."""
+        """Obtains the instantaneous state of the associated ROS system.
+
+        Raises
+        ------
+        ConflictingTypeException
+            If more than one type is detected for topics, services, or actions
```
Comment (ChrisTimperley) [inline src/roswire/ros2/state.py:124]: slight tweak: `is detected for a given topic, service, or action.`

**Review by ChrisTimperley — APPROVED** [review]
👍🏻 LGTM

**Pull request merged**